src/wayland/handlers/pointer_constraints.rs

```diff
@@ -27,11 +27,78 @@ impl PointerConstraintsHandler for State {
 
     fn cursor_position_hint(
         &mut self,
-        _surface: &WlSurface,
-        _pointer: &PointerHandle<Self>,
-        _location: Point<f64, Logical>,
+        surface: &WlSurface,
+        pointer: &PointerHandle<Self>,
+        location: Point<f64, Logical>,
     ) {
-        // TODO
+        if with_pointer_constraint(surface, pointer, |constraint| {
+            constraint.is_some_and(|c| c.is_active())
+        }) {
+            for (out, set) in &self.common.shell.read().workspaces.sets {
+                for workspace in &set.workspaces {
```

Comment:
Given you don't seem to be using the output, you can just use `.spaces()` on `Shell`:

```suggestion
            for workspace in self.common.shell.read().spaces() {
```